### PR TITLE
Contracts: History Expressions via "old" monad

### LINKED ISSUE #2803: History Expression Support for Function Contracts

### PR DESCRIPTION
**Pull request by pi314mm**
Add support for History Expressions. This allows an ensures clause to evaluate expressions before the function runs, bing them to anonymous variables and use them after calculating the result to be able to compare to the state before the function evaluation.

Resolves #2803

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: library/kani_macros/src/sysroot/contracts/shared.rs
Comment (celinval) [inline library/kani_macros/src/sysroot/contracts/shared.rs:171]: please doc what this is doing... I find that giving an example of the expansion is very helpful.

File: tests/expected/function-contract/history/noclone_fail.expected
Comment (celinval) [inline tests/expected/function-contract/history/noclone_fail.expected:1]: Humm... maybe name this something like `kani_old_<var_name>`, so it's a bit more intuitive to the user what's going on.

Another possibility is to create a dummy function:
```rust
fn remember<T: Copy>(val: T) -> T {
  val
}
```

and change the generated code to be:
```rust
let ptr_renamed = kani::internal::remember(ptr);
```

This way, the error will be that `ptr` does not implement Copy trait.

File: tests/expected/function-contract/history/old_old.expected
Comment (celinval) [inline tests/expected/function-contract/history/old_old.expected:1]: that's a nice error

**Review by celinval — COMMENTED** [review]
I didn't do a thorough review, but so far so good. Thanks!

File: tests/expected/function-contract/history/noclone_fail.expected
Comment (pi314mm) [reply tests/expected/function-contract/history/noclone_fail.expected:1]: The naming is actually because of the renaming of variables within the ensures clause
For this particular example,
```rust
#[kani::ensures(|result| old(ptr).0 + 1 == ptr.0)]
```
becomes
```rust
let ptr_renamed = kani::internal::untracked_deref(&ptr);
let result_kani_internal = function_call(ptr);
kani::assert((|result| old(ptr_renamed).0 + 1 == ptr_renamed.0)(result_kani_internal))
std::mem::forget(ptr_renamed)
```
So the "_renamed" comes from the renaming the ensures statement which is something tangential to this PR, but a plausible solution is provided in #3234. 

And then what this PR is implementing is lifting "old" up to the top
```rust
let ptr_renamed = kani::internal::untracked_deref(&ptr);
let remember_kani_internal_0 = ptr_renamed;
let result_kani_internal = function_call(ptr);
kani::assert((|result| remember_kani_internal_0.0 + 1 == ptr_renamed.0)(result_kani_internal))
std::mem::forget(ptr_renamed)
```
This specific error statement comes from the fact that ownership of `ptr_renamed` gets given to the `remember_kani_internal`, but then we try to access it again via the assert statement which causes the borrow of moved value error at the assert.

This error could be solved by making yet another copy of the `let ptr_for_remember = kani::internal::untracked_deref(&ptr);` which essentially functions as contraction within rust affine's logic. This would allow us to remember non-copyable/non-cloneable data, because the move into the `remember_kani_internal_0` wouldn't clash with borrows within the ensures statement later.

File: tests/expected/function-contract/history/old_old.expected
Comment (pi314mm) [reply tests/expected/function-contract/history/old_old.expected:1]: credit for that goes to #2993

File: library/kani_macros/src/sysroot/contracts/shared.rs
Comment (pi314mm) [reply library/kani_macros/src/sysroot/contracts/shared.rs:182]: I now added various documentation in shared.rs and library/kani/src/contracts.rs, and added an example of the expansion within mod.rs

**Comment by pi314mm** [comment]
@JustusAdam could you please review this change as it is closely related (same `old` visitor) to your previous implementation #2993?

File: library/kani_macros/src/sysroot/contracts/shared.rs
Comment (JustusAdam) [inline library/kani_macros/src/sysroot/contracts/shared.rs:152]: Are these remember variables ever exposed to the user? If not I would prefer to use a short hash of the entire function body as a prefix or suffix rather than a count.

File: tests/expected/function-contract/history/block.expected
Comment (JustusAdam) [inline tests/expected/function-contract/history/block.expected:1]: These should generally also list the important (i.e. specifically relevant to the particular test case) as expected. The reason is that then we also ensure that the test case is not vacuous.

File: tests/expected/function-contract/history/noclone_pass.rs
Comment (JustusAdam) [inline tests/expected/function-contract/history/noclone_pass.rs:1]: I would prefer `copy_pass` as a name here

File: library/kani/src/contracts.rs
Comment (JustusAdam) [inline library/kani/src/contracts.rs:234]: A very important point here is to mention whether the absence of side effects here is _required_ or whether it is _checked_. I believe the latter is the case, because CBMC will deny any side-effect use in contracts. 

Might even warrant a test case.

File: library/kani/src/contracts.rs
Comment (JustusAdam) [inline library/kani/src/contracts.rs:229]:
```suggestion
//! ## History Expressions
```

File: library/kani/src/contracts.rs
Comment (JustusAdam) [inline library/kani/src/contracts.rs:231]: "history computations" is not defined or a well known term. I would suggest 

 ```suggestion
//! Additionally, an ensures clause is allowed to refer to the state of the function arguments before function execution and perform simple computations on them
```

**Review by JustusAdam — CHANGES_REQUESTED** [review]
Thanks for your contribution. I recognize some of the code here 😄 

I'm a bit confused as to the terminology of an "old _monad_". To me this terminology suggests that there is some way to produce values in a category representing the state before the function execution (lets call this category O) as opposed to some "current" category (lets call it T). I would then expect some way to apply transformations to values in O and some way to extract O -> T in a specific context. For example I would expect I could write code similar to this 

```rs
fn check_args(x: u32, y: u32) -> O<bool> {
     let x0 : O<u32> = old(x);
     x0.and_then(|x1| O::pure(x1 + y < 9))
}

// Implicitly in the `O` monad
#[ensures(check_args(x, y))]
fn foo(x: u32, y: u32) { ... }
```

I guess what I'm trying to say is I don't see what the natural transformations of this monad would be. 

Aside: One more reason this is likely not a monad, nesting in a monad is always well defined, whereas we deny expressions of the form `old(old(...))`.

**Comment by pi314mm** [comment]
Thank you for the questions about monads! The old monad links the "language of the past" to the "language of the present" so it relates to your explanation via categories.

It is important to understand there is an underlying theory and the theory is related to but not fully implemented by this particular implementation. While we are trying to express the "old monad," implementing the full generality of the monad is rather difficult, so we focus on a particular usage of the monad.

In particular, if we have an ensures statement like
```rust
#[kani::ensures(old(*ptr)+1==*ptr)]
```
Then we comprehend this as syntax for the statement (not within rust)
```
bind (*ptr : O(u32)) (|remember : u32| remember + 1 == *ptr)
```
Which then gets compiled to (within rust)
```rust
let remember = *ptr;
let result = ...;
kani::assert(remember + 1 == *ptr)
```
This means that the underlying principle of the monad is there, but external syntax appears to be less like a monad because otherwise it would be too difficult to implement, and the user most likely only cares about this particular construction of prefixing all the `bind` operations.

As for nested old, such as `old(old(*ptr)+*ptr)`, it is reasonable to interpret this as compiling to
```rust
let remember_1 = *ptr;
let remember_0 = remember_1 + *ptr;
let result = ...;
...
```
so the restriction is just a matter of there not being implementation support for this kind of statement rather than the theory itself. It is not particularly useful to implement this because we claim that there should be no effectful computation within the contracts, so you can substitute the `remember_1` into the second line without worrying about the effects. Hence, we opt for simply restricting this behavior instead of implementing it.

File: library/kani/src/contracts.rs
Comment (pi314mm) [reply library/kani/src/contracts.rs:234]: Required and not enforced by kani. Added a new test case in that fails due to side effects within the old computation. You may still be right for other kani contracts, but it isn't checked for old.

File: tests/expected/function-contract/history/noclone_fail.expected
Comment (pi314mm) [reply tests/expected/function-contract/history/noclone_fail.expected]: decided to do a shuffling around of code so that it uses the name `ptr` which is tracked by rust and gives a nicer error message

File: tests/expected/function-contract/history/block.expected
Comment (feliperodri) [inline tests/expected/function-contract/history/block.expected:5]: Add new line.

File: tests/expected/function-contract/history/function_call.expected
Comment (feliperodri) [inline tests/expected/function-contract/history/function_call.expected:5]: Add new line.

File: tests/expected/function-contract/history/copy_pass.rs
Comment (feliperodri) [inline tests/expected/function-contract/history/copy_pass.rs:3]: It would be nice if test cases had a little description about what is intended behavior to be tested here. Sometimes the test description contains an error message that is very obvious, which might ultimately making this extra description unnecessary. So, you can use your judgment where to add it.

File: tests/expected/function-contract/history/old_old.expected
Comment (feliperodri) [inline tests/expected/function-contract/history/old_old.expected:2]: Add new line.

File: tests/expected/function-contract/history/side_effect.expected
Comment (feliperodri) [inline tests/expected/function-contract/history/side_effect.expected:2]: Add new line.

File: tests/expected/function-contract/history/simple_fail.expected
Comment (feliperodri) [inline tests/expected/function-contract/history/simple_fail.expected:7]: Add new line.

File: tests/expected/function-contract/history/simple_pass.expected
Comment (feliperodri) [inline tests/expected/function-contract/history/simple_pass.expected:5]: Add new line.

File: tests/expected/function-contract/history/simple_fail.rs
Comment (feliperodri) [inline tests/expected/function-contract/history/simple_fail.rs:5]: We can replace `|result|` with simply `||` if we don't even have a return value. Maybe it makes the case cleaner.

File: tests/expected/function-contract/history/simple_pass_twice.expected
Comment (feliperodri) [inline tests/expected/function-contract/history/simple_pass_twice.expected:9]: Add new line.

File: tests/expected/function-contract/history/simple_pass_twice.rs
Comment (feliperodri) [inline tests/expected/function-contract/history/simple_pass_twice.rs:13]: What about test cases that use `stub_verified`? These are important because `old` can help us to write more precise ensure clauses, which would restrict more the value havoced during replacement. Could you add some of those as well? Also, what happen if I use old without an modifies clause? Would this lead to an error?

File: library/kani/src/contracts.rs
Comment (feliperodri) [inline library/kani/src/contracts.rs:249]: Could you also update our function contracts [RFC](https://github.com/model-checking/kani/blob/main/rfc/src/rfcs/0009-function-contracts.md#history-expressions) to reflect this new implementation? Let's also add a comment in the issue #2747 to remember to include a section in our documentation about history variables with examples.

File: library/kani_macros/src/sysroot/contracts/helpers.rs
Comment (feliperodri) [inline library/kani_macros/src/sysroot/contracts/helpers.rs:288]: Add a short comment/documentation to this macro rule. Take a look at this [example](https://github.com/model-checking/kani/blob/9ee1d296de22f363b399df7a38d56c5796bfed1b/library/kani/src/tuple.rs#L13).

File: library/kani_macros/src/sysroot/contracts/mod.rs
Comment (feliperodri) [inline library/kani_macros/src/sysroot/contracts/mod.rs:243]:
```suggestion
//! state. Each occurrence of `old` is lifted, so is is necessary that
//! each lifted occurrence is closed with respect to the function arguments.
```

File: library/kani_macros/src/sysroot/contracts/shared.rs
Comment (feliperodri) [reply library/kani_macros/src/sysroot/contracts/shared.rs:184]: @JustusAdam Why would that be preferable in this case?

File: library/kani_macros/src/sysroot/contracts/shared.rs
Comment (feliperodri) [inline library/kani_macros/src/sysroot/contracts/shared.rs:214]:
```suggestion
/// When a `#[kani::ensures(|result|expr)]` is expanded, this function is called on
/// with `build_ensures(|result|expr, remember_count)` where `remember_count` is the total number of
/// `remember_kani_internal_` variables that exist before building this ensures statement.
/// This function goes through the expr and extracts out all the `old` expressions and creates a sequence
/// of statements that instantiate these expressions as `let remember_kani_internal_x = old_expr;` with
/// `x` starting the `remember_count` and incrementing from there. This is returned as the first return
/// parameter along with changing all the variables to `_renamed`. The second parameter is the closing of
/// all the unsafe argument copies. The third return parameter is the expression formed by passing in the
/// result variable into the input closure and changing all the variables to `_renamed`.
```

File: library/kani_macros/src/sysroot/contracts/shared.rs
Comment (feliperodri) [inline library/kani_macros/src/sysroot/contracts/shared.rs:249]:
```suggestion
    /// You are provided with the expression that is the first argument of the
    /// `old()` call. You may modify it as you see fit. The return value
    /// indicates whether the entire `old()` call should be replaced by the
    /// (potentially altered) first argument.
    ///
    /// The second argument is the span of the original `old` expression.
    ///
    /// The third argument is the number of remember variables that have already been
    /// instantiated in the surrounding environment.
    ///
    /// The fourth argument is a collection of all the expressions that need to be lifted
    /// into the past environment as new remember variables.
```

File: library/kani_macros/src/sysroot/contracts/shared.rs
Comment (feliperodri) [inline library/kani_macros/src/sysroot/contracts/shared.rs:297]: Let's add a string description for each assertion to provide a better error message.

File: library/kani_macros/src/sysroot/contracts/shared.rs
Comment (feliperodri) [inline library/kani_macros/src/sysroot/contracts/shared.rs:339]:
```suggestion
        // The index of the `remembers_exprs` is offset by the `remember_count` of the surrounding environment
```

**Review by feliperodri — CHANGES_REQUESTED** [review]
Only a few minor changes (mostly documentation) before we merge. Thanks for this!!

**Comment by feliperodri** [comment]
@JustusAdam could you give another review?

File: library/kani_macros/src/sysroot/contracts/shared.rs
Comment (JustusAdam) [reply library/kani_macros/src/sysroot/contracts/shared.rs:184]: Counting is brittle, plus code that matches over syntax (e.g. this function) are effort to maintain if the AST format changes. We should already have a routine that calculated a unique hash (it may even be stored in the generators already). Using a hash would completely eliminate this syntax matching function.

**Review by JustusAdam — APPROVED** [review]
Looks good to me otherwise

**Review by celinval — CHANGES_REQUESTED** [review]
Can you please add a few `ui/` tests where the user misuses the API, for example:
  1. User tries to pass an object of type `T: !Copy` to `old`.
  2. `old` has no argument.
  3. `old` argument is not an expression.
  4. User tries to pass `result` to `old`.

File: tests/expected/function-contract/history/ui/statement.expected
Comment (celinval) [inline tests/expected/function-contract/history/ui/statement.expected:1]: What is the error message that the user would see in this case?

**Review by celinval — COMMENTED** [review]
Thanks for adding the `ui` tests. I think we need to work on these error messages though. I'm OK with making it a follow up PR, but I'll leave it up to you and @feliperodri.

File: tests/expected/function-contract/history/ui/statement.expected
Comment (pi314mm) [reply tests/expected/function-contract/history/ui/statement.expected:1]:
```
error[E0433]: failed to resolve: maybe a missing crate `core`?
 --> tests/expected/function-contract/history/ui/statement.rs:5:41
  |
5 | #[kani::ensures(|result| old(let x = 1+1;) old(*ptr + 1) == *ptr)]
  |                                         ^ maybe a missing crate `core`?
  ```
  Not sure if it's the error message we actually want, but it's also very malformated so it's really hard to get a good error message.

File: library/kani_macros/src/sysroot/contracts/shared.rs
Comment (pi314mm) [reply library/kani_macros/src/sysroot/contracts/shared.rs]: Now using syn::Expr's Hash trait to hash the contents of `old(expr)`. If there is a hash collision, it would be because the exprs are identical, which means we can use the same variable for both expressions.

File: tests/expected/function-contract/history/ui/noncopy_ignore.expected
Comment (feliperodri) [inline tests/expected/function-contract/history/ui/noncopy_ignore.expected:1]: New line.

File: tests/expected/function-contract/history/ui/no_args.rs
Comment (feliperodri) [inline tests/expected/function-contract/history/ui/no_args.rs:5]: Add a description of what you are trying to test here.

File: tests/expected/function-contract/history/ui/statement.expected
Comment (feliperodri) [reply tests/expected/function-contract/history/ui/statement.expected:1]: @celinval should we create a GitHub issue to improve these error messages on a separate PR?

**Review by feliperodri — APPROVED** [review]

File: tests/expected/function-contract/history/ui/statement.expected
Comment (celinval) [reply tests/expected/function-contract/history/ui/statement.expected:1]: Sure. I'm OK with this being a follow up PR, but we definitely need to improve these messages.

Comment (pi314mm) [reply tests/expected/function-contract/history/ui/statement.expected:1]: Issue mentioned in #3273

File: library/kani_macros/src/sysroot/contracts/helpers.rs
Comment (pi314mm) [reply library/kani_macros/src/sysroot/contracts/helpers.rs:288]: Issue mentioned in #3273

File: library/kani_macros/src/sysroot/contracts/shared.rs
Comment (pi314mm) [reply library/kani_macros/src/sysroot/contracts/shared.rs:261]: Issue mentioned in #3273

**Pull request merged**
